### PR TITLE
create property only marketplace

### PR DESCRIPTION
**Pull request by eskarlat**
Още няма тести за проперти утре ще ги направя, ми герми грешките:

`  Contract: Property
    creating property proxy
      1) "before each" hook for "should get the owner of the first contract"
    upgrade property contract
      2) "before each" hook for "should upgrade contract from owner"
    create new Property
      3) "before each" hook for "should create new Property"

  0 passing (430ms)
  3 failing

  1) Contract: Property creating property proxy "before each" hook for "should get the owner of the first contract":
     Error: The contract code couldn't be stored, please check your gas amount.
      at Object.callback (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:329221:46)
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:39618:25
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:331159:9
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:314196:9
      at XMLHttpRequest.request.onreadystatechange (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:315621:13)
      at XMLHttpRequestEventTarget.dispatchEvent (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70159:18)
      at XMLHttpRequest._setReadyState (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70449:12)
      at XMLHttpRequest._onHttpResponseEnd (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70604:12)
      at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70564:24)
      at endReadableNT (_stream_readable.js:1054:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  2) Contract: Property upgrade property contract "before each" hook for "should upgrade contract from owner":
     Error: The contract code couldn't be stored, please check your gas amount.
      at Object.callback (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:329221:46)
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:39618:25
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:331159:9
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:314196:9
      at XMLHttpRequest.request.onreadystatechange (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:315621:13)
      at XMLHttpRequestEventTarget.dispatchEvent (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70159:18)
      at XMLHttpRequest._setReadyState (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70449:12)
      at XMLHttpRequest._onHttpResponseEnd (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70604:12)
      at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70564:24)
      at endReadableNT (_stream_readable.js:1054:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  3) Contract: Property create new Property "before each" hook for "should create new Property":
     Error: The contract code couldn't be stored, please check your gas amount.
      at Object.callback (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:329221:46)
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:39618:25
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:331159:9
      at /usr/local/lib/node_modules/truffle/build/cli.bundled.js:314196:9
      at XMLHttpRequest.request.onreadystatechange (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:315621:13)
      at XMLHttpRequestEventTarget.dispatchEvent (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70159:18)
      at XMLHttpRequest._setReadyState (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70449:12)
      at XMLHttpRequest._onHttpResponseEnd (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70604:12)
      at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/truffle/build/cli.bundled.js:70564:24)
      at endReadableNT (_stream_readable.js:1054:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)`

### CODE REVIEW

File: contracts/Property/Property.sol [hunk 1]
```diff
@@ -90,6 +91,7 @@ contract Property is IProperty, OwnableUpgradeableImplementation, Pausable {
         bool _isInstantBooking
 		) public onlyInactive(_propertyId) whenNotPaused returns(bool success)
 	{
+        require(Marketplace(msg.sender).isMarketPlace());
         require(_marketplaceId != "");
```
Comment (thcrnk) [inline contracts/Property/Property.sol]: Here you could add require checking if marketplaceId is approved to be one hundred percents sure that the id is correct

File: contracts/Property/Property.sol [hunk 2]
```diff
@@ -79,18 +83,26 @@ contract Property is IProperty, OwnableUpgradeableImplementation, Pausable {
         );
     }
 
-    function createProperty(
+    function create(
```
Comment (Perseverance) [inline contracts/Property/Property.sol]: Add host address as param and put it as hostAddress

File: migrations/2_marketplace_contracts.js [hunk 3]
```diff
@@ -1,9 +1,19 @@
...
+    let PropertyInstance = await Property.deployed();
+    await deployer.deploy(PropertyProxy, PropertyInstance.address);
+    await PropertyInstance.init();
+
     await deployer.deploy(Marketplace);
     let MarketplaceInstance = await Marketplace.deployed();
-    await deployer.deploy(MarketplaceProxy, MarketplaceInstance.address);   
-    await MarketplaceInstance.init();
+    await deployer.deploy(MarketplaceProxy, MarketplaceInstance.address);
+    // console.log("HEEEEEEEEEEEEEEEEEEE*****"+PropertyInstance.address);
```
Comment (Perseverance) [inline migrations/2_marketplace_contracts.js]: Remove this log

File: test/Marketplace.js [hunk 4]
```diff
@@ -29,37 +41,56 @@ contract('Marketplace', function (accounts) {
...
       marketplaceContract = await IMarketplace.at(proxy.address);
-      await marketplaceContract.init();
-		});
 
-		it("should get the owner of the first contract", async function () {
-			const owner = await marketplaceContract.getOwner();
-			assert.strictEqual(owner, _owner, "The owner is not set correctly");
-		});
-	});
+      impl3 = await Property.new();
```
Comment (Perseverance) [inline test/Marketplace.js]: Rename this to propertyImpl everywhere

File: test/Marketplace.js [hunk 5]
```diff
@@ -29,37 +41,56 @@ contract('Marketplace', function (accounts) {
...
+      proxy3 = await PropertyProxy.new(impl3.address);
```
Comment (Perseverance) [inline test/Marketplace.js]: Rename this to propertyProxy

File: test/Marketplace.js [hunk 6]
```diff
@@ -29,37 +41,56 @@ contract('Marketplace', function (accounts) {
...
+  const _workingDayPrice = '1000000000000000000';
+  const _nonWorkingDayPrice = '2000000000000000000';
+  const _cleaningFee = '100000000000000000';
+  const _refundPercent = '80';
+  const _daysBeforeStartForRefund = '10';
+  const _isInstantBooking = true;
+
+  describe("creating marketplace proxy", () => {
+    beforeEach(async function () {
+      impl = await Marketplace.new();
```
Comment (Perseverance) [inline test/Marketplace.js]: Rename this to marketplaceImpl everywhere

File: test/Marketplace.js [hunk 7]
```diff
@@ -29,37 +41,56 @@ contract('Marketplace', function (accounts) {
...
+      proxy = await MarketplaceProxy.new(impl.address);
```
Comment (Perseverance) [inline test/Marketplace.js]: Rename this to marketplaceProxy

File: test/Marketplace.js [hunk 8]
```diff
@@ -594,56 +661,124 @@ contract('Marketplace', function (accounts) {
...
+        _refundPercent,
+        _cleaningFee,
+        _daysBeforeStartForRefund,
+        _isInstantBooking, {
+          from: _owner
+        }
+      );
+
+      assert.isTrue(Boolean(result.receipt.status), "The propertiesCount creation was not successful");
+    });
```
Comment (Perseverance) [inline test/Marketplace.js]: Add more tests

File: contracts/Marketplace/IMarketplace.sol [hunk 9]
```diff
@@ -10,6 +10,13 @@ contract IMarketplace is IOwnableUpgradeableImplementation, IPausable {
 	event LogApproveMarketplace(bytes32 marketplaceId);
 	event LogRejectMarketplace(bytes32 marketplaceId);
 	event LogChangeApprovalPolicy(bool isApprovalPolicyActive);
+    event LogCreatePropertyFromMarketplace(bytes32 propertyId, address hostAddress);
```
Comment (thcrnk) [inline contracts/Marketplace/IMarketplace.sol]: Add marketplaceId to the event

File: contracts/Marketplace/Marketplace.sol [hunk 10]
```diff
@@ -26,9 +29,15 @@ contract Marketplace is IMarketplace, OwnableUpgradeableImplementation, Pausable
 	event LogApproveMarketplace(bytes32 marketplaceId);
 	event LogRejectMarketplace(bytes32 marketplaceId);
 	event LogChangeApprovalPolicy(bool isApprovalPolicyActive);
+    event LogCreatePropertyFromMarketplace(bytes32 propertyId, address hostAddress);
```
Comment (thcrnk) [inline contracts/Marketplace/Marketplace.sol]: Add marketplaceId to the event

File: test/Property.js [hunk 11]
```diff
@@ -196,23 +89,15 @@ contract('Property', function (accounts) {
...
-        _workingDayPrice,
-        _nonWorkingDayPrice,
-        _cleaningFee,
-        _refundPercent,
-        _daysBeforeStartForRefund,
-        _isInstantBooking, {
-          from: _propertyHost
-        }
-      );
+    it("should throw if trying to create Property when paused", async function () {
```
Comment (thcrnk) [inline test/Property.js]: This test will fail even if the contract is not paused because there is no Marketplace.

File: test/Marketplace.js [hunk 12]
```diff
@@ -1,22 +1,33 @@
 const web3 = require("web3");
+//Marketplace
 const MarketplaceProxy = artifacts.require("./Marketplace/MarketplaceProxy.sol");
 const Marketplace = artifacts.require("./Marketplace/Marketplace.sol");
 const IMarketplace = artifacts.require("./Marketplace/IMarketplace.sol");
+
+//Property
```
Comment (thcrnk) [inline test/Marketplace.js]: It clear here that those requires are for Property - comment not needed

File: test/Marketplace.js [hunk 13]
```diff
@@ -1,22 +1,33 @@
 const web3 = require("web3");
+//Marketplace
```
Comment (thcrnk) [inline test/Marketplace.js]: It clear here that those requires are for Marketplace - comment not needed

File: test/Property.js [hunk 14]
```diff
@@ -71,120 +81,18 @@ contract('Property', function (accounts) {
...
-          from: _propertyHost
-        }
-      );
-
-      assert.isTrue(Boolean(result.receipt.status), "The Property creation was not successful");
-
-      let propertiesCount = await PropertyContract.propertiesCount();
-      assert(propertiesCount.eq(1), "The Propertys count was not correct");
 
+      marketplaceImpl = await Marketplace.new();
```
Comment (thcrnk) [inline test/Property.js]: marketplace implementation should not be in this BeforeEach. Move it to paused test and add approve Marketplace.

File: contracts/Property/Property.sol [hunk 15]
```diff
@@ -90,10 +95,16 @@ contract Property is IProperty, OwnableUpgradeableImplementation, Pausable {
         bool _isInstantBooking
 		) public onlyInactive(_propertyId) whenNotPaused returns(bool success)
```
Comment (Perseverance) [inline contracts/Property/Property.sol]: Add unit test for onlyInactive

File: contracts/Property/Property.sol [hunk 16]
```diff
@@ -90,10 +95,16 @@ contract Property is IProperty, OwnableUpgradeableImplementation, Pausable {
...
 	{
+        
         require(_marketplaceId != "");
+        require(_hostAddress != address(0));
```
Comment (Perseverance) [inline contracts/Property/Property.sol]: Add unit test for host 0x0

File: contracts/Property/Property.sol [hunk 17]
```diff
@@ -90,10 +95,16 @@ contract Property is IProperty, OwnableUpgradeableImplementation, Pausable {
         bool _isInstantBooking
 		) public onlyInactive(_propertyId) whenNotPaused returns(bool success)
 	{
+        
         require(_marketplaceId != "");
```
Comment (Perseverance) [inline contracts/Property/Property.sol]: Add unit test for marketplaceId = ""

File: contracts/Property/Property.sol [hunk 18]
```diff
@@ -90,10 +95,16 @@ contract Property is IProperty, OwnableUpgradeableImplementation, Pausable {
...
+        require(_hostAddress != address(0));
 
+        MarketplaceContract = IMarketplace(msg.sender);
+        require(MarketplaceContract.isMarketplace());
+        require(MarketplaceContract.isApprovedMarketplace(_marketplaceId));
```
Comment (Perseverance) [inline contracts/Property/Property.sol]: Add unit test for marketplace that is not approved

Comment (eskarlat) [reply contracts/Property/Property.sol]: this test already has in test: 

it("should throw when marketplace is reject", async function () {

      await marketplaceContract.rejectMarketplace(
        _marketplaceId, {
          from: _owner
        }
      );

      await expectThrow(marketplaceContract.createProperty(
        _propertyId,
        _marketplaceId,
        _workingDayPrice,
        _nonWorkingDayPrice,
        _cleaningFee,
        _refundPercent,
        _daysBeforeStartForRefund,
        _isInstantBooking, {
          from: _propertyHost
        }
      ));
    });

File: contracts/Property/Property.sol (hunk 16 above)
Comment (eskarlat) [reply contracts/Property/Property.sol]: Нали msg.sender  не може да бъде 0х0 защото _hostAddress това е msg.sender 
PropertyContract.create(
            _propertyId,
            _marketplaceId, 
            msg.sender,
            _workingDayPrice,
            _nonWorkingDayPrice,
            _cleaningFee,
            _refundPercent,
            _daysBeforeStartForRefund,
            _isInstantBooking
        );

File: contracts/Property/Property.sol (hunk 15 above)
Comment (eskarlat) [reply contracts/Property/Property.sol]: пак имаме този тест:
it("should throw if the same PropertyId is used twice", async function () {
      await marketplaceContract.createProperty(
        _propertyId,
        _marketplaceId,
        _workingDayPrice,
        _nonWorkingDayPrice,
        _cleaningFee,
        _refundPercent,
        _daysBeforeStartForRefund,
        _isInstantBooking, {
          from: _propertyHost
        }
      );

      await expectThrow(marketplaceContract.createProperty(
        _propertyId,
        _marketplaceId,
        _workingDayPrice,
        _nonWorkingDayPrice,
        _cleaningFee,
        _refundPercent,
        _daysBeforeStartForRefund,
        _isInstantBooking, {
          from: _propertyHost
        }
      ));
    });

File: contracts/Property/Property.sol (hunk 17 above)
Comment (eskarlat) [reply contracts/Property/Property.sol]: Marketplace.js

it("should throw if trying to create Property with empty marketplaceId", async function () {
      await expectThrow(marketplaceContract.createProperty(
        _propertyId,
        "",
        _workingDayPrice,
        _nonWorkingDayPrice,
        _cleaningFee,
        _refundPercent,
        _daysBeforeStartForRefund,
        _isInstantBooking, {
          from: _propertyHost
        }
      ));
    });

File: contracts/Property/Property.sol (hunk 15 above)
Comment (Perseverance) [reply contracts/Property/Property.sol]: Move this to Property.js then ;)

File: contracts/Property/Property.sol (hunk 16 above)
Comment (Perseverance) [reply contracts/Property/Property.sol]: Correct

File: contracts/Property/Property.sol (hunk 18 above)
Comment (Perseverance) [reply contracts/Property/Property.sol]: Move this to Property.js

**Pull request merged**